src/compiler/checker.ts

```diff
@@ -21983,8 +21984,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 }
             }
             if (reportErrors) {
-                // Elaborate only if we can find a best matching type in the target union
-                const bestMatchingType = getBestMatchingType(source, target, isRelatedTo);
+                // Elaborate only if we can find a best matching type in the target
+                const bestMatchingType = getBestMatchingType(source, target, /*matchSingleOverlappy*/ true, isRelatedTo);

```

Comment:
Returning a single overlappy type is required here to make the non-elementwise union-based elaboration to elaborate about the "last" element in the union, see the "recovered" errors in this commit: https://github.com/microsoft/TypeScript/pull/57537/commits/e9d136a2b89a7a7c7f535dcdf0f5bfadce9396c0